### PR TITLE
rework JSON conversions for transaction metadata

### PR DESCRIPTION
**Pull request by KtorZ**
a. Encode JSON as CBOR considering only the following "optimizations"
  b. Hexadecimal sequences starting with 0x are encoded as CBOR bytestring.
  c. JSON keys that are numbers are encoded as CBOR numbers.
  d. CBOR is shown as JSON when possible, or as a string representing JSON-encoded data when not (i.e. when map keys aren't numbers or strings).
  e. CBOR bytestrings are represented as hexadecimal strings prefixed with 0x

  - we expect are that we can round trip any JSON without constraints (modulo null and boolean values)
  - for CBOR, we should be able to round trip a subset with some simple-ish constraints.

### CODE REVIEW

File: cardano-api/test/Test/Cardano/Api/MetaData.hs [hunk 1]
```diff
@@ -6,34 +6,37 @@ module Test.Cardano.Api.MetaData
...
 import qualified Hedgehog.Range as Range
 
 -- This test is fragile to changes in the way Metadata is converted to/from JSON.
 prop_round_trip_json_metadatum :: Property
 prop_round_trip_json_metadatum = do
   Hedgehog.withTests 1000 . Hedgehog.property $ do
-    md <- Hedgehog.forAll genMetaData
-    Hedgehog.tripping md jsonFromMetadata jsonToMetadata
+    json <- jsonFromMetadata <$> Hedgehog.forAll genMetaData
+    Hedgehog.tripping json identity (fmap jsonFromMetadata . jsonToMetadata)
```
Comment (erikd) [inline cardano-api/test/Test/Cardano/Api/MetaData.hs:30]: What? Why is `identity` being used there? That defeats the prurpose of the `tripping` function's ability to pretty print failed test cases.

File: cardano-api/test/Test/Cardano/Api/MetaData.hs [hunk 2]
```diff
@@ -44,43 +47,24 @@ genTxMetadataValue =
...
-              , (1, genFlatTxMetadataValue)
-              ]
-  -- Keys need to be unique and sorted.
-  -- They need to be unque because they are inserted into a HashMap and they
-  -- need to be sorted so they round trip correctly.
-  mapM (\k -> (k,) <$> genTxMetadataValue) (List.sort $ List.nub keys)
+genText :: Gen Text
+genText = Gen.choice
+  [ Gen.ensure (not . Text.isPrefixOf bytesPrefix)
+      (Text.pack <$> Gen.list (Range.linear 0 64) Gen.alphaNum)
```
Comment (erikd) [inline cardano-api/test/Test/Cardano/Api/MetaData.hs:61]: This looks suspiciously like it is generating a subset of the real world so that is can pass round tripping.

**Review by erikd — COMMENTED** [review]
These comments are just my first impressions. I still want to pull this branch and play with it a bit.

**Comment by erikd** [comment]
Reading the PR comments and and looking at the code, I still have no clue what this PR is actually intended to achieve or why.

File: cardano-api/src/Cardano/Api/MetaData.hs [hunk 3]
```diff
@@ -60,26 +62,31 @@ jsonFromMetadataValue md =
...
+      Aeson.String $ bytesPrefix <> Text.decodeUtf8 (Base16.encode bs)
+
+-- JSON only allows keys to be strings. So, there really are two cases:
+--
+-- - Either the keys of the CBOR map are indeed strings and we render them as such
+-- - Or they are something else, and we render them as a serialized JSON string
+--
+-- So, for metadata coming from JSON in the first place, this will be pretty much invisible.
+-- And for more elaborated metadata crafted by other mean (via a CBOR library for instance),
+-- the key in the JSON-rendered metadata will look at bit funky.
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply cardano-api/src/Cardano/Api/MetaData.hs:151]: `s/other mean/other means/`

File: cardano-api/src/Cardano/Api/MetaData.hs [hunk 4]
```diff
@@ -123,49 +129,41 @@ jsonToMetadataValue  av =
         case Scientific.floatingOrInteger sci :: Either Double Integer of
           Left  n -> Left (ConversionErrNumberNotInteger n)
           Right n -> Right (Api.TxMetaNumber n)
-      Aeson.String txt -> jsonToMetadataString txt
+      Aeson.String txt ->
+        case Text.stripPrefix bytesPrefix txt of
+          Nothing  -> jsonToMetadataString txt
+          Just hex -> case Base16.decode (Text.encodeUtf8 hex) of
+            (bytes, "") -> jsonToMetadataBytes bytes
+            (_, err)    -> Left $ ConversionErrBadBytes err
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply cardano-api/src/Cardano/Api/MetaData.hs:214]: Seems like this kind of check should be done by a separate linter than be rejected here.

**Comment by erikd** [comment]
I still do not understand what **specific** problem this PR is trying to address and how it is trying to address it.

File: cardano-api/test/Test/Cardano/Api/MetaData.hs (hunk 1 above)
Comment (rvl) [reply cardano-api/test/Test/Cardano/Api/MetaData.hs]: Fixed - we need to convert the generated metadata to JSON before round-trip testing to "wash out" ambiguous metadatums. For example, `MD.S "0x"` becomes `MD.B ""` after a round-trip, causing the property to fail. I don't really like this though.

File: cardano-api/test/Test/Cardano/Api/MetaData.hs (hunk 2 above)
Comment (rvl) [reply cardano-api/test/Test/Cardano/Api/MetaData.hs]: Fixed.

**Comment by rvl** [comment]
The specific problem - according to me - is that if cardano-wallet uses these json conversion functions, wallet users will submit their JSON metadata with a transaction, and may well find that the JSON is different when they view their metadata in the tx history. This could break naive apps.

Also if users submit JSON metadata like:

```
"metadata": { "42": [["ada", 1], ["ada", 2]] }
```

it will appear in the cardano explorer as:

```
"metadata": { "42": { "ada": 1 /* or 2 */ } }
```

**Comment by erikd** [comment]
Ok, that is a problem statement, but how does this PR address it?

File: cardano-api/test/Test/Cardano/Api/MetaData.hs [hunk 5]
```diff
@@ -1,39 +1,54 @@
...
 prop_round_trip_json_metadatum :: Property
 prop_round_trip_json_metadatum = do
   Hedgehog.withTests 1000 . Hedgehog.property $ do
     md <- Hedgehog.forAll genMetaData
-    Hedgehog.tripping md jsonFromMetadata jsonToMetadata
-
--- -----------------------------------------------------------------------------
+    -- pre-wash metadata to remove ambiguous metadatums before
+    -- testing the round-trip property.
+    let md' = fromRight md $ jsonToMetadata $ jsonFromMetadata md
```
Comment (rvl) [inline cardano-api/test/Test/Cardano/Api/MetaData.hs:35]: On second thoughts, this is going to make the round-trip test much less effective...

**Comment by erikd** [comment]
This needs a lot more thought. Rushing to wrong solutions would be a bad idea.

File: cardano-api/src/Cardano/Api/MetaData.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (KtorZ) [reply cardano-api/src/Cardano/Api/MetaData.hs]: It is a little funky indeed. Cf the slack conversation.

It gives however some nice debugging and inspection capabilities which is in the end, what we really care about for the Json view. 

Another option would be to use a Json schema that can represent any CBOR data, like a diagnostic view but it'll be much more verbose.

In practice, most metadata would originally come from Json so it won't be _that_ weird.

File: cardano-api/src/Cardano/Api/MetaData.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (KtorZ) [reply cardano-api/src/Cardano/Api/MetaData.hs:214]: @intricate yes it is fine. Metadata aren't meant to be a place of self expression but rather a compact way to embed custom information in transactions. So rather than free-form text and verbose sentence, one should put the minimal information possible in metadata and have more verbose details stored off chain.

**Comment by KtorZ** [comment]
@erikd I believe the problem is well stated on the Slack thread I opened some days ago. It is as Rodney pointed out above:

We want metadata to round-trip **when coming from JSON**.

So you're right that the generator only generate a subset of possible values, because this is exactly what we want. Ideally, the generator should generate a Json Value and round-trip from that.

That PR addresses the problem because it makes sure that regardless of the what Json is given, the user get the exact same one after converting to CBOR and to JSON back. It wasn't the case before because of the special transformation done on lists and maps.

The trade-off is that we have some funky keys when translating CBOR maps that have non string keys. That is perhaps a little ugly, yet see the slack thread for rationale.

File: cardano-api/test/Test/Cardano/Api/MetaData.hs (hunk 5 above)
Comment (KtorZ) [reply cardano-api/test/Test/Cardano/Api/MetaData.hs:35]: Ideally, I'd rather generate arbitrary JSON values and see that they roundtrip fine using the metadata conversion function, it'll better capture what is the actual property we want. By re-using jsonToMetadata and jsonFromMetadata to construct the input I am afraid we are indeed creating a bias in the generator which will fail to capture some edge cases.

**Comment by erikd** [comment]
The Slack thread was huge, and making sense of it is far from trivial.

Please summarize, preferably giving examples of:

* How the new serialisation differs from the old
* What the new serialisation handles that the old one does not.

**Review by erikd — CHANGES_REQUESTED** [review]
I *think* I know what the problem is that you are trying to solve and I *think* I have a better solutuon, but until you actually tell me what the problem is, I can't actually tell.

File: cardano-api/src/Cardano/Api/MetaData.hs [hunk 6]
```diff
@@ -123,49 +130,41 @@ jsonToMetadataValue  av =
...
+jsonToMetadataBytes :: ByteString -> Either MetaDataJsonConversionError TxMetadataValue
+jsonToMetadataBytes bytes =
+    if BS.length bytes > maxStringSize
+      then Left ConversionErrLongerThan64Bytes
+      else Right (Api.TxMetaBytes bytes)
+
+-- | JSON strings that are base16 encoded and prefixed with 'bytesPrefix' will
+-- be encoded as CBOR bytestrings.
+bytesPrefix :: Text
+bytesPrefix = "0x"
```
Comment (erikd) [inline cardano-api/src/Cardano/Api/MetaData.hs:167]: This is a bad idea.

**Review by erikd — CHANGES_REQUESTED** [review]

**Comment by rvl** [comment]
How about this JSON encoding scheme? It's round-trippable in both directions.

```yaml
x-transactionMetadata: &transactionMetadata
  type: object
  nullable: true
  additionalProperties:
    $ref: "#/components/schemas/TransactionMetadatum"
  propertyNames:
    pattern: '^[0-9]+$'
  example:
    0: "cardano"
    1: 14
    2: { "hex": "2512a00e9653fe49a44a5886202e24d77eeb998f" }
    3: { "list": [14, 42, 1337] }
    4: { "map": [["key", "value"], ["14", 42]] }

components:
  schemas:
    TransactionMetadatum: &TransactionMetadatum
      oneOf:
        - title: String
          type: string
        - title: ByteString
          type: object
          required:
            - hex
          properties:
            hex:
              type: string
              pattern: "^[0-9a-fA-F]$"
        - title: Integer
          type: integer
        - title: List
          type: object
          required:
            - list
          properties:
            list:
              type: array
              items:
                $ref: "#/components/schemas/TransactionMetadatum"
        - title: Map
          type: object
          required:
            - map
          properties:
            map:
              type: array
              items:
                type: array
                minItems: 2
                maxItems: 2
                items:
                  $ref: "#/components/schemas/TransactionMetadatum"

```

**Comment by intricate** [comment]
@rvl: I'd much prefer something like that. 

Of course, we do lose the ability to specify certain kinds of tx metadata via JSON such as arrays of 2-element arrays (as those would be interpreted as tx metadata maps) and objects with a `"hex"` key that aren't intended to represent byte strings, but I think that's a fair trade-off.

**Comment by intricate** [comment]
@rvl: I'd much prefer something like that. 

@erikd @dcoutts: Thoughts?

File: cardano-api/src/Cardano/Api/MetaData.hs (hunk 6 above)
Comment (KtorZ) [reply cardano-api/src/Cardano/Api/MetaData.hs:167]: May I ask why "it's a bad idea". 

Also, when "requesting changes" on a PR, it's better to tell the author what changes are expected.

**Comment by KtorZ** [comment]
@rvl that would solve problems indeed. May I suggest to call "hex" simply "bytes" for transparency and consistency with others?

**Comment by rvl** [comment]
> May I suggest to call "hex" simply "bytes"

Yes I think that would be better. I was also wondering whether to add "string" and "integer" tags to string and integer values, also for the sake of consistency.

**Comment by KtorZ** [comment]
So, final proposal / outcome from recent Slack conversations, we are going for an explicit meta-schema that encodes in JSON the data-types of the on-chain metadata as Rodney's suggested, but including an explicit name for numbers and strings as well which gives us the following OpenAPI schema:

```yaml
x-transactionMetadata: &transactionMetadata
  type: object
  nullable: true
  additionalProperties:
    $ref: "#/components/schemas/TransactionMetadatum"
  propertyNames:
    pattern: '^[0-9]+$'
  example:
    0: { "string": "cardano" }
    1: { "int": 14 }
    2: { "bytes": "2512a00e9653fe49a44a5886202e24d77eeb998f" }
    3: { "list": [{ "int": 14 }, { "int": 42 }, { "int": 1337 }] }
    4: { "map": [[{"string": "key"}, {"string": "value"}], [{"string": "14"}, {"int": 42}]] }

components:
  schemas:
    TransactionMetadatum: &TransactionMetadatum
      oneOf:
        - title: String
          type: object
          required:
            - string
          properties:
            string: 
              type: string
              maxLength: 64

        - title: Int
          type: object
          required:
            - int 
          properties:
            int: 
              type: integer

        - title: ByteString
          type: object
          required:
            - bytes
          properties:
            bytes:
              type: string
              pattern: "^[0-9a-fA-F]$"
              maxLength: 128

        - title: List
          type: object
          required:
            - list
          properties:
            list:
              type: array
              items:
                $ref: "#/components/schemas/TransactionMetadatum"

        - title: Map
          type: object
          required:
            - map
          properties:
            map:
              type: array
              items:
                type: array
                minItems: 2
                maxItems: 2
                items:
                  $ref: "#/components/schemas/TransactionMetadatum"
```

**Comment by dcoutts** [comment]
We can have more than one choice for rendering in JSON of the underlying metadata capability.

This PR takes the approach of allowing any JSON to be encoded to metadata, but it does not expose the full capability of the underlying metadata structure. This has the advantage of not needing a JSON schema, since any JSON is permitted.

The alternative is to expose the full capability of the underlying metadata structure. This needs a JSON schema.

Different use cases need different things. It's totally ok for the API to support both of the above approaches.

File: cardano-api/src/Cardano/Api/MetaData.hs (hunk 6 above)
Comment (erikd) [reply cardano-api/src/Cardano/Api/MetaData.hs:167]: We used the `0x` and it made it difficult to distingush `ByteString` from `Text`. Property testing immediately found all the holes and the only way to fix it was to represent `ByteString`s an object like `{ "bytes" : "ffffff" }`.

**Comment by KtorZ** [comment]
rebased on top of 1.19.1

**Comment by dcoutts** [comment]
Still TODO:
* update the cli
* simplify the schema error reporting data type

**Comment by rooooooooob** [comment]
I don't see any mention of alternate encodings (e.g. strings) for too-big ints (metadata can have u64 uint and u64 nint as it uses CDDL int) in that OpenAPI schema. Granted I am not familiar with it, but I quickly scanned the docs and saw 2 optional format specifiers for `int32` and `int64` for the `integer` used there but even `int64` would only include half of the possible values of a CDDL `int` since those are almost like a 65-bit integer. The code in this PR also mentions 64 bit integers.

**Comment by dcoutts** [comment]
> I don't see any mention of alternate encodings (e.g. strings) for too-big ints (metadata can have u64 uint and u64 nint as it uses CDDL int) in that OpenAPI schema.

Can you elaborate on what you mean, since it's not clear to me. Yes the tx metadata can have up to 64bit negative and positive integers. The code checks these limits when converting from JSON.

There is no alternative in the tx metadata for integers bigger than 64bit. That is simply the maximum.

**Comment by rooooooooob** [comment]
tl;dr: If we have a large number like 2^64 - 1 in the metadata CBOR, how does this convert to JSON?

The 65 bit comment was if we were using signed 64-bit to represent both then we would effectively be losing one bit of information, no? `int = uint / nint` which is almost basically a 65-bit signed int since it's unsigned 64 in both signs. My concern was just when it came to how we would be representing numbers big enough to be in a 64-bit uint/nint but not enough to be in a 64-bit signed integer. For example representing 2^63 + n, 0 <= n < 2^63 or the same but negative.

```haskell
   [ TxMetadataNumberOutOfRange n
    |    n >         fromIntegral (maxBound :: Word64)
      || n < negate (fromIntegral (maxBound :: Word64))
    ]
```

Are you referring to this? This is only checking the metadata limits, not related to JSON.

My concern was not how the Haskell/CBOR part is doing this, but how we are handling this in JSON. I thought that JSON integers are signed 32-bytes. And beyond that, even what I saw in the OpenAPI docs it just mentioned an optional signed 64-bit integer format tag, which still doesn't represent half (although a half that is not likely to ever be used as they'd be very large/very small) of the possible values of a CDDL `int`. I didn't see that optional format tag used so I wasn't sure if it was even using that or how we would be representing >32-bit numbers in JSON when we convert from the haskell metadata which allows up to 64-bit positive or negative unsigned.

**Comment by dcoutts** [comment]
JSON itself supports arbitrary precision numbers in scientific format, so we're not limited there at all.

Some JSON libs in some programming languages have tighter limitations, but that's the problem on their side. We can produce and consume valid JSON with large integers no problem at all.

File: cardano-api/src/Cardano/Api/MetaData.hs (hunk 6 above)
Comment (dcoutts) [reply cardano-api/src/Cardano/Api/MetaData.hs:352]: :sunglasses: property testing demonstrating things working fine.

**Review by dcoutts — APPROVED** [review]
This is totally cheating now, for me to review what is now my own PR :grin: (having taken it over).

**Comment by dcoutts** [comment]
bors merge

**Review by Jimbo4350 — COMMENTED** [review]
LGTM!

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 7]
```diff
@@ -907,6 +909,25 @@ pPoolMetaDataFile =
       <> Opt.completer (Opt.bashCompleter "file")
       )
 
+pTxMetadataJsonSchema :: Parser TxMetadataJsonSchema
+pTxMetadataJsonSchema =
+    (  Opt.flag' ()
+        (  Opt.long "--json-metadata-no-schema"
```
Comment (ArturWieczorek) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:915]: I guess that two hyphens in `Opt.long "--json-metadata-no-schema"` and on line 922 `Opt.long "--json-metadata-detailed-schema"` are not needed.
